### PR TITLE
Structure_Engine: Add associated Engine methods for BarRebarIntent and PanelRebarIntent

### PR DESCRIPTION
**Pull request by peterjamesnugent**
### NOTE: Depends on 

https://github.com/BHoM/BHoM/pull/1255
   
### Test files

https://burohappold.sharepoint.com/:u:/r/sites/BHoM/02_Current/12_Scripts/01_Issue/BHoM/Versioning_Toolkit/%23168-ConcreteSectionUpgrade.gh?csf=1&web=1&e=YqA4Bo

### Changelog

- Refactor existing engine methods that make use of  `IBarReinforcement`, `LongitudinalReinforcement` or `TransverseReinforcement`
- Update `Create` methods that have inputs for `MinimumCover` and `List<IBarReinforcement>`
- Added `Create` methods for `BarRebarIntent`, `PanelRebarIntent`, `PanelReinforcement` and `ReinforcementRegion`
- Added associated `IsNull` methods for the new objects

### Additional comments

Versioning still needs to be added.

### CODE REVIEW

**Comment by peterjamesnugent** [comment]
@BHoMBot check compliance

**Comment by peterjamesnugent** [comment]
@BHoMBot check compliance

File: Structure_Engine/Create/SectionProperties/ConcreteSection/ConcreteCircularSection.cs
Comment (IsakNaslundBh) [inline Structure_Engine/Create/SectionProperties/ConcreteSection/ConcreteCircularSection.cs:50]: Not necessarily against this change, but it is not needed really, unless we want to.

Can simply have the methods generating a RebarIntent for you and applying that to the section.

You are working closer to the workflows for this though, so can leave it to you to say what works best script-wise.

File: Structure_Engine/Create/SectionProperties/RebarIntent/BarRebarIntent.cs
Comment (IsakNaslundBh) [inline Structure_Engine/Create/SectionProperties/RebarIntent/BarRebarIntent.cs:49]: Method not needed as it is a simple property assigner, so would just remove it.

If kept, would be carefull about nullchecking, as could be cases were you for example would like to just setup intents with minimum covers, and then add the reinforcement by some other mechanism.

File: Structure_Engine/Create/SectionProperties/RebarIntent/PanelRebarIntent.cs
Comment (IsakNaslundBh) [inline Structure_Engine/Create/SectionProperties/RebarIntent/PanelRebarIntent.cs:42]: Method not needed as it is a simple property assigner, so would just remove it.

File: Structure_Engine/Create/SectionProperties/RebarIntent/PanelReinforcement.cs
Comment (IsakNaslundBh) [inline Structure_Engine/Create/SectionProperties/RebarIntent/PanelReinforcement.cs:49]: Method not needed as it is a simple property assigner, so would just remove it.

File: Structure_Engine/Create/SectionProperties/RebarIntent/ReinforcementRegion.cs
Comment (IsakNaslundBh) [inline Structure_Engine/Create/SectionProperties/RebarIntent/ReinforcementRegion.cs:45]: Method not needed as it is a simple property assigner, so would just remove it.

Also, if kept, would for sure get rid of the nullchecks, as I think we agreed null property == use panel default (full region and/or panel coordinatesystem)

File: Structure_Engine/Modify/AddRebarIntent.cs
Comment (IsakNaslundBh) [inline Structure_Engine/Modify/AddRebarIntent.cs:45]: Would revert most of the changes in this method, as the intention was to be able to add more reinforcement to a section that already potentially had some reinforcement on it.

Would keep signature and most of the descriptions as before, but change the behaviour to instead add reinforcement to the RebarIntent obejct owned by the concrete section.

File: Structure_Engine/Modify/SetRebarIntent.cs
Comment (IsakNaslundBh) [inline Structure_Engine/Modify/SetRebarIntent.cs:45]: As above. WOuld keep method signature intact, but change the behaviour to setting the reinforcement of the inner RebarIntent

File: Structure_Engine/Query/ReinforcementLayout.cs
Comment (IsakNaslundBh) [inline Structure_Engine/Query/ReinforcementLayout.cs:404]: Check the inner list for null as well `section.RebarIntent.BarReinforcement == null` before checking the count

File: Structure_Engine/Query/ReinforcementLayout.cs
Comment (IsakNaslundBh) [inline Structure_Engine/Query/ReinforcementLayout.cs:414]: Check the inner list for null as well `section.RebarIntent.BarReinforcement == null` before checking the count

File: Structure_Engine/Query/ReinforcementTransitionPoints.cs
Comment (IsakNaslundBh) [inline Structure_Engine/Query/ReinforcementTransitionPoints.cs:48]: Check the inner list for null as well `concreteSection.RebarIntent.BarReinforcement == null` before checking the count

**Review by IsakNaslundBh — CHANGES_REQUESTED** [review]
Some initial comments

File: Structure_Engine/Create/SectionProperties/RebarIntent/BarRebarIntent.cs
Comment (IsakNaslundBh) [reply Structure_Engine/Create/SectionProperties/RebarIntent/BarRebarIntent.cs:49]: Also, before you remove, make sure to bring over any of the descriptions to the oM!

File: Structure_Engine/Create/SectionProperties/ConcreteSection/ConcreteCircularSection.cs
Comment (peterjamesnugent) [reply Structure_Engine/Create/SectionProperties/ConcreteSection/ConcreteCircularSection.cs:50]: I guess it depends on our discussion here https://github.com/BHoM/BHoM/pull/1255#discussion_r664513720 about where the `MinimumCover` is stored. 

From a workflow point of view, it's down to how much we thinking `BarRebarIntent` objects will be used as a storage mechanism outside of `ConcreteSection`, and then when assigning it you'd have to explode it for the `Create` method.

**Review by pawelbaran — COMMENTED** [review]
Nothing more to add code-wise besides what mentioned by @IsakNaslundBh, however won't we need a bit of versioning here?

**Comment by peterjamesnugent** [comment]
@BHoMBot check compliance
@BHoMBot check versioning
@BHoMBot check null-handling

**Comment by peterjamesnugent** [comment]
@BHoMBot check compliance

**Comment by peterjamesnugent** [comment]
@BHoMBot check compliance

**Comment by peterjamesnugent** [comment]
@BHoMBot check compliance

**Comment by peterjamesnugent** [comment]
@FraserGreenroyd check compliance

**Comment by peterjamesnugent** [comment]
@BHoMBot check compliance

**Review by pawelbaran — CHANGES_REQUESTED** [review]
Sorry for coming out of the blue, but shouldn't `Create` methods for Reinforcement be migrated to _Structure_Engine/Create/Reinforcement_ folder?

**Review by pawelbaran — DISMISSED** [review]
Looks much better to me now 👍 However, I would opt for removal of _Create\Reinforcement\Reinforcement.cs_ altogether instead of moving it around - it looks like some ancient legacy that should not have the right to exist any more.

**Comment by peterjamesnugent** [comment]
@pawelbaran agreed - going to do it in a separate PR to not bloat this one.

**Comment by pawelbaran** [comment]
Oke, unblocking the PR then - happy about the changes, but would be good to have a final green light from @IsakNaslundBh 👍

**Review by pawelbaran — COMMENTED** [review]
LGTM, final word to @IsakNaslundBh

File: Structure_Engine/Query/MaterialComposition.cs
Comment (IsakNaslundBh) [inline Structure_Engine/Query/MaterialComposition.cs:91]: This is kind of nit-picking, but the concrete sections material composition is not necessarily homogenous. If it contains reinforcement, it will try to grab it, and if we add a description, think it should cover that.

File: Structure_Engine/Query/ReinforcementLayout.cs
Comment (IsakNaslundBh) [inline Structure_Engine/Query/ReinforcementLayout.cs:404]: This will raise an error message that was not there before (reinforcement list being empty).

Was this intentional?

File: Structure_Engine/Query/ReinforcementLayout.cs
Comment (IsakNaslundBh) [inline Structure_Engine/Query/ReinforcementLayout.cs:414]: as above

File: Structure_Engine/Query/ReinforcementTransitionPoints.cs
Comment (IsakNaslundBh) [inline Structure_Engine/Query/ReinforcementTransitionPoints.cs:48]: as above

**Review by IsakNaslundBh — COMMENTED** [review]
One nit-picky description update and a question regarding methods that now will start raising errors that previously did not

File: Structure_Engine/Query/MaterialComposition.cs
Comment (peterjamesnugent) [reply Structure_Engine/Query/MaterialComposition.cs:91]: Amended.

File: Structure_Engine/Query/ReinforcementLayout.cs
Comment (peterjamesnugent) [reply Structure_Engine/Query/ReinforcementLayout.cs:404]: Will only raise an error message if `BarRebarIntent` is provided and the `Reinforcement` is `null` or empty which would be an edge case and worthy of an error.

The `null` check for the `BarRebarIntent` will capture any default `ConcreteSection` and no error message will be provided as  an empty List<LongitudinalReinforcement>` will be returned.

**Review by IsakNaslundBh — APPROVED** [review]
looks good to me now!

**Comment by IsakNaslundBh** [comment]
@BHoMBot check complaince

**Comment by IsakNaslundBh** [comment]
@BHoMBot check compliance

**Comment by IsakNaslundBh** [comment]
@BHoMBot check core

**Comment by IsakNaslundBh** [comment]
@BHoMBot check serialisation
@BHoMBot check null-handling

**Comment by IsakNaslundBh** [comment]
@BHoMBot check ready-to-merge

**Pull request merged**
